File: src/main/java/com/enderryno/nuclearcraft/services/ItemRegister.java
package com.enderryno.nuclearcraft.services;

import com.enderryno.nuclearcraft.classes.PluginConfiguration;
import com.enderryno.nuclearcraft.enums.ConfigurationStorages;
import com.enderryno.nuclearcraft.classes.CustomItem;
import com.enderryno.nuclearcraft.interfaces.PluginItem;
import com.enderryno.nuclearcraft.enums.ItemBehaviour;
import com.enderryno.nuclearcraft.exceptions.ItemNotRegisteredException;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;

/**
 * This class handles the registration of the custom items
 * All the registered items are stored and returned in a form of a HashMap(id, CustomItem)
 *
 * To initialize create an instance of ItemRegister and execute the method registerItems, it will automatically
 * load up correctly the custom items
 *
 */
public class ItemRegister {

    /**
     * Static member to access the registered items
     */
    private static HashMap<Integer, PluginItem> registeredItems = null;

    private JavaPlugin pluginInstance = null;

    /**
     *
     * @param pluginInstance - The instance of the plugin
     */
    public ItemRegister(JavaPlugin pluginInstance) {
        this.pluginInstance = pluginInstance;
    }

    /**
     *
     * @return boolean - True if all items are registered, false if some error occurred.
     */
    public boolean registerItems () {

        registeredItems = new HashMap<>();
        /* REGISTER THE ITEMS COMING FROM THE CONFIG */
        PluginConfiguration itemConfigurator = new PluginConfiguration(this.pluginInstance, ConfigurationStorages.items);
        FileConfiguration itemConfig = itemConfigurator.getConfig();

        if (itemConfig == null) {
            return false;
        }

        if (itemConfig.getList("enabled-items") == null) {
            return false;
        }

        itemConfig.getList("enabled-items").forEach(item -> {

            PluginItem customItem = new CustomItem();

            int itemId = itemConfig.getInt(item + ".item-id");
            if (itemId == 0 || registeredItems.get(itemId) != null) return;

            String itemName = itemConfig.getString(item + ".item-name");
            String itemDescription = itemConfig.getString(item + ".item-description");
            String itemMinecraftId = itemConfig.getString(item + ".item-minecraft-id");

            int itemDataModelId = itemConfig.getInt(item + ".item-data-model-id");
            int itemCustomBlockId = itemConfig.getInt(item + ".custom-block-id");

            int itemStackSize = itemConfig.getInt(item + ".max-stack-size");

            boolean itemEquipable = itemConfig.getBoolean(item + ".is-equipable");
            boolean itemUsable = itemConfig.getBoolean(item + ".is-usable");
            boolean itemTransportable = itemConfig.getBoolean(item + ".is-transportable");

            customItem.setID(itemId);
            customItem.setName(itemName);
            customItem.setDescription(itemDescription);
            customItem.setMinecraftId(itemMinecraftId);

            customItem.setModelId(itemDataModelId);
            customItem.setCustomBlockId(itemCustomBlockId);

            customItem.setStackSize(itemStackSize);

            customItem.setEquipable(itemEquipable);
            customItem.setUsable(itemUsable);
            customItem.setTransportable(itemTransportable);

            String behaviour = itemConfig.getString(item + ".behaviour");
            if (behaviour == null) {
                behaviour = "generic";
            }

            customItem.setBehaviour(ItemBehaviour.fromString(behaviour));



            registeredItems.put(customItem.getID(), customItem);

        });



        itemConfigurator.saveConfig();
        return true;
    }



    /* Registered items getter */
    public static HashMap<Integer, PluginItem> getRegisteredItems() throws ItemNotRegisteredException {
        if (registeredItems == null) {
            throw new ItemNotRegisteredException("Item were not registered! Verify the initialization");
        }


        return registeredItems;
    }


}

File: src/main/java/com/enderryno/nuclearcraft/services/ItemRegister.kt
package com.enderryno.nuclearcraft.services

import com.enderryno.nuclearcraft.NuclearCraft
import com.enderryno.nuclearcraft.classes.CustomItem
import com.enderryno.nuclearcraft.classes.PluginConfiguration
import com.enderryno.nuclearcraft.enums.ConfigurationStorages
import com.enderryno.nuclearcraft.enums.ItemBehaviour
import com.enderryno.nuclearcraft.exceptions.ItemNotRegisteredException
import com.enderryno.nuclearcraft.interfaces.PluginItem
import org.bukkit.plugin.java.JavaPlugin

/**
 * This class handles the registration of the custom items
 * All the registered items are stored and returned in a form of a HashMap(id, CustomItem)
 *
 * To initialize create an instance of ItemRegister and execute the method registerItems, it will automatically
 * load up correctly the custom items
 *
 */
class ItemRegister() {
    private val pluginInstance: JavaPlugin = JavaPlugin.getPlugin(NuclearCraft::class.java)


    /**
     *
     * @return boolean - True if all items are registered, false if some error occurred.
     */
    fun registerItems(): Boolean {
        registeredItems = HashMap()
        /* REGISTER THE ITEMS COMING FROM THE CONFIG */
        val itemConfigurator = PluginConfiguration(pluginInstance, ConfigurationStorages.Items)
        val itemConfig = itemConfigurator.config ?: return false
        if (itemConfig.getList("enabled-items") == null) {
            return false
        }
        itemConfig.getList("enabled-items")!!.forEach { item: Any? ->
            val itemId = item.toString()
            if (registeredItems!![itemId] != null) return@forEach

            val itemName = itemConfig.getString("$item.item-name") ?: throw ItemNotRegisteredException(itemId)
            val itemDescription = itemConfig.getString("$item.item-description")
            val itemMinecraftId = itemConfig.getString("$item.item-minecraft-id") ?: throw ItemNotRegisteredException(itemId)
            val itemDataModelId = itemConfig.getInt("$item.item-data-model-id")
            val itemCustomBlockId = itemConfig.getString("$item.custom-block-id")

            val itemStackSize = itemConfig.getInt("$item.max-stack-size")
            val itemEquipable = itemConfig.getBoolean("$item.is-equipable")
            val itemUsable = itemConfig.getBoolean("$item.is-usable")
            val itemTransportable = itemConfig.getBoolean("$item.is-transportable")
            val itemDroppable = itemConfig.getBoolean("$item.is-droppable")


            var behaviourName = itemConfig.getString("$item.behaviour")
            if (behaviourName == null) {
                behaviourName = "generic"
            }
            val behaviourValue = ItemBehaviour.fromString(behaviourName) ?: throw IllegalArgumentException("Behaviour $behaviourName is not valid")

            val customItem: PluginItem = CustomItem(
                    id = itemId,
                    name = itemName,
                    description = itemDescription,
                    minecraftId = itemMinecraftId,
                    modelId = itemDataModelId,
                    customBlockId = itemCustomBlockId,

                    isEquipable = itemEquipable,
                    isUsable = itemUsable,
                    isTransportable = itemTransportable,
                    stackSize = itemStackSize,
                    isDroppable = itemDroppable,
                    behaviour = behaviourValue
            )
            NuclearCraft.Companion.Logger.info("Registered item $itemId")
            registeredItems!![customItem.id] = customItem
        }
        itemConfigurator.saveConfig()
        return true
    }

    companion object {
        /**
         * Static member to access the registered items
         */
        private var registeredItems: HashMap<String?, PluginItem?>? = null

        /* Registered items getter */
        @Throws(ItemNotRegisteredException::class)
        fun getRegisteredItems(): HashMap<String?, PluginItem?>? {
            if (registeredItems == null) {
                throw ItemNotRegisteredException("Item were not registered! Verify the initialization")
            }
            return registeredItems
        }
    }
}
